File: backend/app/usecases/gps_analysis.py
"""ユースケース: GPSバッチ分析 — 交差点での一時停止判定とリルート処理。"""

from geopy.distance import geodesic

from app.domain.models import GpsPoint, IntersectionResult, Route
from app.domain.ports import GpsRepository, RouteRepository, RoutingService, TripRepository


class GpsAnalysisResult:
    __slots__ = ("saved", "intersection_results", "rerouted")

    def __init__(
        self,
        saved: int,
        intersection_results: list[IntersectionResult],
        rerouted: bool = False,
    ) -> None:
        self.saved = saved
        self.intersection_results = intersection_results
        self.rerouted = rerouted


class GpsAnalysisUseCase:
    def __init__(
        self,
        gps_repo: GpsRepository,
        route_repo: RouteRepository,
        trip_repo: TripRepository,
        routing_service: RoutingService,
        radius_m: float = 15.0,
        speed_threshold: float = 3.0,
        off_route_threshold_m: float = 50.0,
    ) -> None:
        self._gps_repo = gps_repo
        self._route_repo = route_repo
        self._trip_repo = trip_repo
        self._routing_service = routing_service
        self._radius_m = radius_m
        self._speed_threshold = speed_threshold
        self._off_route_threshold_m = off_route_threshold_m

    async def execute(
        self, trip_id: str, points: list[GpsPoint],
    ) -> GpsAnalysisResult:
        self._gps_repo.append_points(trip_id, points)

        rerouted = False
        route = self._route_repo.get_route(trip_id)

        # 経路逸脱時にリルート
        if route and points:
            last_point = points[-1]
            if self._is_off_route(last_point, route):
                trip = self._trip_repo.get(trip_id)
                destination = None
                if trip and trip.destination_lat is not None and trip.destination_lng is not None:
                    destination = (trip.destination_lat, trip.destination_lng)
                if destination:
                    new_route = await self._routing_service.get_bicycle_route(
                        (last_point.lat, last_point.lng), destination,
                    )
                    new_results = [
                        IntersectionResult(intersection=ix)
                        for ix in new_route.intersections
                    ]
                    self._route_repo.save_route(trip_id, new_route)
                    self._route_repo.save_intersection_results(trip_id, new_results)
                    route = new_route
                    rerouted = True

        # 交差点での一時停止を判定
        results = self._route_repo.get_intersection_results(trip_id)
        for point in points:
            if point.accuracy_m > 20:
                continue
            for result in results:
                if result.stopped:
                    continue
                dist = geodesic(
                    (point.lat, point.lng),
                    (result.intersection.lat, result.intersection.lng),
                ).meters
                if dist <= self._radius_m:
                    if result.min_speed_kmh is None or point.speed_kmh < result.min_speed_kmh:
                        result.min_speed_kmh = point.speed_kmh
                    if point.speed_kmh < self._speed_threshold:
                        result.stopped = True

        self._route_repo.save_intersection_results(trip_id, results)

        return GpsAnalysisResult(
            saved=len(points),
            intersection_results=results,
            rerouted=rerouted,
        )

    def _is_off_route(self, point: GpsPoint, route: Route) -> bool:
        """ポイントがルートジオメトリのいずれかのセグメントから離れすぎていないか判定。"""
        min_dist = float("inf")
        for lat, lng in route.geometry:
            dist = geodesic((point.lat, point.lng), (lat, lng)).meters
            if dist < min_dist:
                min_dist = dist
            if min_dist <= self._off_route_threshold_m:
                return False
        return min_dist > self._off_route_threshold_m

File: backend-go/internal/usecase/gps_analysis.go
package usecase

import (
	"btd/internal/domain"
	"context"
	"log/slog"
	"time"
)

// GpsAnalysisUseCase はGPSバッチデータを受信し，交差点での一時停止判定と経路逸脱検知を行うユースケース．
// フロントエンドから5秒間隔で送信されるGPSポイントを処理する．
type GpsAnalysisUseCase struct {
	gpsRepo            domain.GpsRepository   // GPSポイントの永続化
	routeRepo          domain.RouteRepository // ルート・交差点結果の読み書き
	tripRepo           domain.TripRepository  // 目的地の取得（リルート時に必要）
	routingService     domain.RoutingService  // 再ルーティング（経路逸脱時）
	radiusM            float64                // 交差点圏内と判定する半径（m）
	speedThreshold     float64                // 一時停止と判定する速度閾値（km/h）
	offRouteThresholdM float64                // 経路逸脱と判定する距離（m）
	accuracyThresholdM float64                // GPS精度フィルタ閾値（m）．これを超えるポイントは無視
}

func NewGpsAnalysisUseCase(
	gpsRepo domain.GpsRepository,
	routeRepo domain.RouteRepository,
	tripRepo domain.TripRepository,
	routingService domain.RoutingService,
	radiusM, speedThreshold, offRouteThresholdM, accuracyThresholdM float64,
) *GpsAnalysisUseCase {
	return &GpsAnalysisUseCase{
		gpsRepo:            gpsRepo,
		routeRepo:          routeRepo,
		tripRepo:           tripRepo,
		routingService:     routingService,
		radiusM:            radiusM,
		speedThreshold:     speedThreshold,
		offRouteThresholdM: offRouteThresholdM,
		accuracyThresholdM: accuracyThresholdM,
	}
}

// Execute はGPSポイントを保存し，経路逸脱チェック→一時停止判定を実行する．
// 処理結果として保存件数，交差点判定結果，リルート有無を返す．
func (uc *GpsAnalysisUseCase) Execute(ctx context.Context, tripID string, points []domain.GpsPoint) (*domain.GpsAnalysisResult, error) {
	tTotal := time.Now()

	if err := uc.gpsRepo.AppendPoints(tripID, points); err != nil {
		return nil, err
	}

	rerouted := false
	route, err := uc.routeRepo.GetRoute(tripID)
	if err != nil {
		return nil, err
	}

	if route != nil && len(points) > 0 {
		lastPoint := points[len(points)-1]
		if uc.isOffRoute(lastPoint, route) {
			trip, err := uc.tripRepo.Get(tripID)
			if err != nil {
				return nil, err
			}
			if trip != nil && trip.DestinationLat != nil && trip.DestinationLng != nil {
				dest := domain.LatLng{Lat: *trip.DestinationLat, Lng: *trip.DestinationLng}
				origin := domain.LatLng{Lat: lastPoint.Lat, Lng: lastPoint.Lng}

				// リルート前の交差点結果を取得し，通過済みのものを保持する
				oldResults, err := uc.routeRepo.GetIntersectionResults(tripID)
				if err != nil {
					return nil, err
				}
				var pastEncountered []*domain.IntersectionResult
				for _, r := range oldResults {
					if r.Stopped || r.MinSpeedKmh != nil {
						pastEncountered = append(pastEncountered, r)
					}
				}

				// 現在位置→目的地で新ルートを取得
				newRoute, err := uc.routingService.GetBicycleRoute(ctx, origin, dest)
				if err != nil {
					return nil, err
				}
				// 新ルートの交差点結果を初期化
				newResults := make([]*domain.IntersectionResult, len(newRoute.Intersections))
				for i, ix := range newRoute.Intersections {
					newResults[i] = &domain.IntersectionResult{Intersection: ix}
				}

				// 旧ルートの通過済み交差点 + 新ルートの交差点を結合し，indexを振り直す
				combined := make([]*domain.IntersectionResult, 0, len(pastEncountered)+len(newResults))
				combined = append(combined, pastEncountered...)
				combined = append(combined, newResults...)
				for i, r := range combined {
					r.Intersection.Index = i
				}

				if err := uc.routeRepo.SaveRoute(tripID, newRoute); err != nil {
					return nil, err
				}
				if err := uc.routeRepo.SaveIntersectionResults(tripID, combined); err != nil {
					return nil, err
				}
				rerouted = true
			}
		}
	}

	tStopDetect := time.Now()
	results, err := uc.routeRepo.GetIntersectionResults(tripID)
	if err != nil {
		return nil, err
	}

	// 速度ポインタ用のバッファを事前確保（ループ内のヒープアロケーションを回避）
	speedBuf := make([]float64, len(results))
	for i, r := range results {
		if r.MinSpeedKmh != nil {
			speedBuf[i] = *r.MinSpeedKmh
		} else {
			speedBuf[i] = -1 // 未設定を示すセンチネル値
		}
	}

	// 全GPSポイント × 全交差点 で判定
	for _, point := range points {
		// GPS精度が低いポイントはスキップ
		if point.AccuracyM > uc.accuracyThresholdM {
			continue
		}
		// 速度が取得できなかったポイントは停止判定に使用しない
		if point.SpeedKmh == nil {
			continue
		}
		speed := *point.SpeedKmh
		for i, result := range results {
			// 既に停止判定済みの交差点はスキップ
			if result.Stopped {
				continue
			}
			// Haversine距離で交差点圏内かを判定
			dist := domain.DistanceM(point.Lat, point.Lng, result.Intersection.Lat, result.Intersection.Lng)
			if dist <= uc.radiusM {
				// 圏内の最小速度を更新
				if speedBuf[i] < 0 || speed < speedBuf[i] {
					speedBuf[i] = speed
					result.MinSpeedKmh = &speedBuf[i]
				}
				// 速度が閾値未満なら一時停止と判定
				if speed < uc.speedThreshold {
					result.Stopped = true
				}
			}
		}
	}

	if err := uc.routeRepo.SaveIntersectionResults(tripID, results); err != nil {
		return nil, err
	}

	slog.Info("GpsAnalysis",
		"total_s", time.Since(tTotal).Seconds(),
		"stop_detect_s", time.Since(tStopDetect).Seconds(),
		"points", len(points),
		"intersections", len(results),
		"rerouted", rerouted,
	)

	return &domain.GpsAnalysisResult{
		Saved:               len(points),
		IntersectionResults: results,
		Rerouted:            rerouted,
	}, nil
}

// isOffRoute は指定GPSポイントがルートから逸脱しているかを判定する．
// ルートジオメトリの全頂点との最小距離が閾値を超えていれば true を返す．
// 閾値以内の頂点が見つかった時点で早期リターンする．
func (uc *GpsAnalysisUseCase) isOffRoute(point domain.GpsPoint, route *domain.Route) bool {
	minDist := float64(1e18)
	for _, ll := range route.Geometry {
		dist := domain.DistanceM(point.Lat, point.Lng, ll.Lat, ll.Lng)
		if dist < minDist {
			minDist = dist
		}
		if minDist <= uc.offRouteThresholdM {
			return false // 経路上にいる
		}
	}
	return minDist > uc.offRouteThresholdM
}
